shards/core/runtime.hpp

```diff
@@ -490,23 +514,23 @@ struct SHMesh : public std::enable_shared_from_this<SHMesh> {
     SHInstanceData data = instanceData;
     data.wire = wire.get();
     data.inputType = shards::deriveTypeInfo(input, data);
-    VisitedWires visitedWires;
-    data.visitedWires = &visitedWires;
-    auto validation = shards::composeWire(
-        wire.get(),
-        [](const Shard *errorShard, SHStringWithLen errorTxt, bool nonfatalWarning, void *userData) {
-          auto blk = const_cast<Shard *>(errorShard);
-          if (!nonfatalWarning) {
-            throw shards::ComposeError(std::string(errorTxt.string, errorTxt.len) +
-                                       ", input shard: " + std::string(blk->name(blk)));
-          } else {
-            SHLOG_INFO("Validation warning: {} input shard: {}", errorTxt, blk->name(blk));
-          }
-        },
-        this, data);
-    shards::arrayFree(validation.exposedInfo);
-    shards::arrayFree(validation.requiredInfo);
-    shards::freeDerivedInfo(data.inputType);
+    CompositionContext privateContext{};
+    data.privateContext = &privateContext;
+    try {
+      auto validation = shards::composeWire(wire.get(), data);
+      shards::arrayFree(validation.exposedInfo);
```

Comment:
You need to free this in the failure case too, since I added some extra checks for traits/stop, or you would have to free it on failure in that code (search for validateWireTraits usage)